libs/s25main/mapGenerator/Resources.cpp

```diff
@@ -177,39 +178,75 @@ namespace rttr { namespace mapGenerator {
     {
         const auto& textures = map.textureMap;
         auto& resources = map.resources;
-        int total = settings.ratioCoal + settings.ratioGold + settings.ratioIron + settings.ratioGranite;
+        struct MineableResourceInfo
+        {
+            // The current "budget", that is, how many many are we over, or under, compared to the desired distribution.
+            int budget = 0;
+            // The ratio is the user defined ratio for this resource.
+            int ratio;
+            // The resource encoding
+            libsiedler2::Resource resource;
+            MineableResourceInfo(int ratio, libsiedler2::Resource resource) : ratio(ratio), resource(resource) {}
+        };
+        auto mRIs = helpers::make_array(MineableResourceInfo(settings.ratioCoal, libsiedler2::R_Coal),
+                                        MineableResourceInfo(settings.ratioGold, libsiedler2::R_Gold),
+                                        MineableResourceInfo(settings.ratioIron, libsiedler2::R_Iron),
+                                        MineableResourceInfo(settings.ratioGranite, libsiedler2::R_Granite));
+        const int total = settings.ratioCoal + settings.ratioGold + settings.ratioIron + settings.ratioGranite;
+
+        // Helper to pick an index [0,4) into the mRIs array that identifies a randomly selected resource we should
+        // place. Note that we do not perform budget adjustments and checks here.
+        auto resourcePicker = [&]() {
+            int ratio = 0;
+            int randomNumber = rnd.RandomValue(1, std::max(1, total));
+            for(unsigned i = 0; i < mRIs.size(); ++i)
+            {
+                ratio += mRIs[i].ratio;
+                if(randomNumber <= ratio)
+                    return i;
+            }
+            RTTR_Assert_Msg(0, "Expected to pick a resource!");
+            return 0u;
+        };
 
         RTTR_FOREACH_PT(MapPoint, map.size)
         {
             if(textures.All(pt, IsMinableMountain))
             {
-                int randomNumber = rnd.RandomValue(1, total);
-                int ratio = settings.ratioGold;
-
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Gold + rnd.RandomValue(0, 8);
+                // No mineable resources wanted.
+                if(total == 0)
                     continue;
-                }
 
-                ratio += settings.ratioCoal;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Coal + rnd.RandomValue(0, 8);
+                // Pick a random resource, -1 indicates none.
```

Comment:
```suggestion
                // Pick a random resource
```

Or alternatively rename the lambda to `pickRandomResourceIdx` then the comment can be removed completely